File: 2018/puzzle3/python/solution.py
from collections import Counter

if __name__ == "__main__":
    with open('../input.txt') as box_ids:
        two_ct = 0
        three_ct = 0

        while True:
            line = box_ids.readline()
            if not line:
                break
            counter = Counter(line)
            if 2 in counter.values():
                two_ct += 1
            if 3 in counter.values():
                three_ct += 1

        print(f'The final checksum is {two_ct * three_ct}')

File: 2018/puzzle3/js/solution.js
const fs = require('fs')

fs.readFile('../input.txt', 'utf8', (err, data) => {
    let boxIds = data.split('\n')
    boxIds.pop()

    let counts = boxIds.reduce((acc, boxId) => {
        let letterCt = boxId
            .split('')
            .reduce((acc, letter) => {
                if (!acc.hasOwnProperty(letter)) acc[letter] = 0;
                acc[letter]++;
                return acc
            }, {})

        if (Object.values(letterCt).includes(2)) acc.twoCt++;
        if (Object.values(letterCt).includes(3)) acc.threeCt++;
        return acc
    }, {twoCt: 0, threeCt: 0})

    console.log(`The final checksum is ${
        counts.twoCt * counts.threeCt
    }`)
})